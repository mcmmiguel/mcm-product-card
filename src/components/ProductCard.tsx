import React from 'react';
import { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import { InitialValuesProps, onChangeArgs, ProductCardHandlersProps, ProductContextProps, ProductProps } from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;


export interface ProductCardProps {
    children?: (args: ProductCardHandlersProps) => JSX.Element;
    product: ProductProps;
    className?: string;
    style?: CSSProperties;
    onChange?: (args: onChangeArgs) => void;
    value?: number;
    initialValues?: InitialValuesProps;
};

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: ProductCardProps) => {

    const { counter, maxCount, increaseBy, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount,
        }}>

            <div className={`${styles.productCard} ${className}`} style={style}>
                {children ? children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,

                    increaseBy,
                    reset,
                }) : null}
                {/* <ProductImage img={product.img} />
                <ProductTitle title={product.title} />
                <ProductButtons counter={counter} increaseBy={increaseBy} /> */}
            </div>
        </Provider>
    );
};