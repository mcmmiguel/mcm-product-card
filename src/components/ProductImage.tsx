import React from 'react';
import { CSSProperties, useContext } from 'react';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface ProductImageProps {
    img?: string;
    className?: string;
    style?: CSSProperties;
}

export const ProductImage = ({ img = '', className, style }: ProductImageProps) => {

    const { product } = useContext(ProductContext);
    let imageToShow: string;

    if (img) {
        imageToShow = img;
    } else if (product.img) {
        imageToShow = product.img;
    } else {
        imageToShow = noImage;
    }

    return (
        <img className={`${styles.productImg} ${className}`} style={style} src={imageToShow} alt="Product" />
    );
};