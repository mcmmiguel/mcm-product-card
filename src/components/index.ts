import { ProductCardHOCProps } from '../interfaces/interfaces';

// export { ProductCard } from './ProductCard';
import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';

export { ProductImage } from './ProductImage';
export { ProductTitle } from './ProductTitle';
export { ProductButtons } from './ProductButtons';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Image: ProductImage,
    Title: ProductTitle,
    Buttons: ProductButtons,
})