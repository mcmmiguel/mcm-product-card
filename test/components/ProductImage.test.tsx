import React from "react";
import renderer from 'react-test-renderer'
import { ProductCard, ProductImage } from "../../src/components";
import { product2 } from "../data/products";


describe('Pruebas en <ProductImage />', () => {

    test('Debe de mostrar correctamente el componente', () => {

        const wrapper = renderer.create(
            <ProductImage img="http://imgdeprueba.jpg" />
        );

        expect(wrapper.toJSON()).toMatchSnapshot();

    });

    test('Debe de mostrar correctamente el componente con la imagen del producto', () => {

        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        );

        expect(wrapper.toJSON()).toMatchSnapshot();

    });

});