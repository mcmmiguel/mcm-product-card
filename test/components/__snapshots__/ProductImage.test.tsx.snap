// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ProductImage /> Debe de mostrar correctamente el componente 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="http://imgdeprueba.jpg"
/>
`;

exports[`Pruebas en <ProductImage /> Debe de mostrar correctamente el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
